File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        #seach for column first
        n = len(matrix)
        m = len(matrix[0])
        low = 0
        high = n - 1
        mid = (low + high) // 2
        row = -1
        while low <= high:
            mid = (low + high) // 2
            if target > matrix[mid][m - 1]:
                low = mid + 1
            else:
                row = mid
                high = mid - 1
        if row == -1:
            return False
        print(row)
        low = 0
        high = m - 1
        while low <= high:
            mid = (low + high) // 2
            if target == matrix[row][mid]:
                return True
            if target > matrix[row][mid]:
                low = mid + 1
            else:
                high = mid - 1
        return False

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();

        //consider 2d array as an plain array from 0 to n*m-1 then binary search
        int left = 0, right = m * n-1;
        int mid;
        while(left <= right){
            mid = left + (right - left)/2;
            int x = mid / n;
            int y = mid % n;
            if(matrix[x][y] == target)
                return true;
            
            if(matrix[x][y] <= target)
                left = mid + 1;
            else
                right = mid - 1;
        }
        return false;

    }
};